openedx_certificates/generators.py

```diff
@@ -0,0 +1,153 @@
+    :return: Username.
+    """
+    return user.profile.name or f"{user.first_name} {user.last_name}"
+
+
+def _register_font(options: dict[str, Any]) -> str:
+    """
+    Register a custom font if specified in options. If not specified, use the default font (Helvetica).
+
+    :param options: A dictionary containing the font.
+    :returns: The font name.
+    """
+    if font := options.get('font'):
+        pdfmetrics.registerFont(TTFont(font, ExternalCertificateAsset.get_asset_by_slug(font)))
+
+    return font or 'Helvetica'
+
+
+def _write_text_on_template(template: any, font: str, username: str, course_name: str, options: dict[str, Any]) -> any:
+    """
+    Prepare a new canvas and write the user and course name onto it.
+
+    :param template: Pdf template.
+    :param font: Font name.
+    :param username: The name of the user to generate the certificate for.
+    :param course_name: The name of the course the learner completed.
+    :param options: A dictionary containing the Y coordinates for name and course name.
+    :returns: A canvas with written data.
+    """
+    template_width, template_height = template.mediabox[2:]
+    pdf_canvas = canvas.Canvas(io.BytesIO(), pagesize=(template_width, template_height))
+    pdf_canvas.setFont(font, 32)
+    # Write the learner name.
+    name_x = (template_width - pdf_canvas.stringWidth(username)) / 2
+    name_y = options.get('name_y', 290)
+    pdf_canvas.drawString(name_x, name_y, username)
+    # Write the course name.
+    pdf_canvas.setFont(font, 28)
+    course_name_x = (template_width - pdf_canvas.stringWidth(course_name)) / 2
+    course_name_y = options.get('course_name_y', 220)
+    pdf_canvas.drawString(course_name_x, course_name_y, course_name)
+    return pdf_canvas
+
+
+def _save_certificate(certificate: PdfWriter, certificate_uuid: UUID) -> str:
+    """
+    Save the final PDF file to BytesIO and upload it using Django default storage.
+
+    :param certificate: Pdf certificate.
+    :param certificate_uuid: The UUID of the certificate.
+    :returns: The URL of the saved certificate.
+    """
+    # Save the final PDF file to BytesIO.
+    output_path = f'external_certificates/{certificate_uuid}.pdf'
+    pdf_bytes = io.BytesIO()
+    certificate.write(pdf_bytes)
+    pdf_bytes.seek(0)  # Rewind to start.
+    # Upload with Django default storage.
+    certificate_file = ContentFile(pdf_bytes.read())
+    # Delete the file if it already exists.
+    if default_storage.exists(output_path):
+        default_storage.delete(output_path)
+    default_storage.save(output_path, certificate_file)
+    if isinstance(default_storage, FileSystemStorage):
+        url = f"{settings.LMS_ROOT_URL}{settings.MEDIA_URL}{output_path}"
+    else:
+        url = default_storage.url(output_path)
+    return url
+
+
+def generate_pdf_certificate(course_id: CourseKey, user: User, certificate_uuid: UUID, options: dict[str, Any]) -> str:
```

Comment:
Will every `generate_` method use this same signature? (for badges, external certs, other types...?)

It seems to be missing the date at which the cert was earned, which may be different from the date when the PDF is generated.